File: week1_baseline/ruby/12_context/test/test_tools_mcp.rb
require "minitest/autorun"

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)

require "boukensha"
require "mud_manager/fake_mud"

# Boukensha as an MCP host.
#
# These spawn real MCP servers as subprocesses, so they exercise the actual
# thing: an agent framework that knows nothing about MUDs acquiring tools at
# runtime. No API key, no billing, no real MUD.
#
# Tool names carry the `tbamud` prefix here. That is a *client-side* policy —
# the daemon still advertises `look` on the wire. If a daemon test ever needs
# updating for the prefix, the prefix leaked across the boundary and that is a
# bug, not a rename.
class TestToolsMcp < Minitest::Test
  DAEMON = File.expand_path("../../../../week0_explore/mud_manager/bin/mud-manager", __dir__)
  CALCULATOR = File.expand_path("support/tiny_mcp_server.rb", __dir__)
  PASSWORD = "swordfish".freeze
  PREFIX = Boukensha::MUD_PREFIX # "tbamud"

  def setup
    skip "daemon not found at #{DAEMON}" unless File.exist?(DAEMON)
    @mud = MudManager::FakeMud.new(password: PASSWORD).start
    @ctx = Boukensha::Context.new(task: Boukensha::Tasks::Player, system: "")
    @registry = Boukensha::Registry.new(@ctx)
    @clients = []
  end

  def teardown
    @clients.each { |c| c.close rescue nil }
    @mud&.stop
  end

  def register_mud(prefix: PREFIX)
    client = Boukensha::Tools::Mcp.register(
      @registry,
      command: RbConfig.ruby,
      args: [DAEMON, "--mcp"],
      env: {
        "MUD_HOST" => "127.0.0.1",
        "MUD_PORT" => @mud.port.to_s,
        "MUD_NAME" => "Gandalf",
        "MUD_PASSWORD" => PASSWORD,
        "BOUKENSHA_DIR" => ""
      },
      prefix: prefix,
      label: "mud"
    )
    @clients << client
    client
  end

  def register_calculator(prefix: "calc")
    client = Boukensha::Tools::Mcp.register(
      @registry, command: RbConfig.ruby, args: [CALCULATOR], prefix: prefix, label: "calc"
    )
    @clients << client
    client
  end

  # ---------- generic host behaviour ----------------------------------------

  def test_registers_a_non_mud_server
    # The point of the whole refactor: Tools::Mcp has no MUD knowledge. Proven
    # by demonstration rather than assertion — if MUD assumptions creep back in,
    # this fails loudly.
    register_calculator

    assert_equal %w[calc__add calc__shout], @ctx.tools.keys.sort
    assert_equal "42.0", @registry.dispatch("calc__add", { "a" => 2, "b" => 40 })
    assert_equal "HELLO", @registry.dispatch("calc__shout", { "text" => "hello" })
  end

  def test_two_servers_coexist
    register_mud
    register_calculator

    assert @ctx.tools.key?("#{PREFIX}__look")
    assert @ctx.tools.key?("calc__add")
    assert_equal 28, @ctx.tools.length # 26 MUD + 2 calculator
  end

  def test_collision_raises_and_names_the_server
    # Registry#tool would silently clobber. Prefixing makes collisions unlikely,
    # not impossible — two entries can share a prefix — so the check stays.
    register_calculator(prefix: "dup")

    err = assert_raises(Boukensha::Tools::Mcp::ToolCollisionError) do
      register_calculator(prefix: "dup")
    end

    assert_match(/dup__add/, err.message)
    assert_match(/prefix/, err.message)
  end

  def test_bare_names_still_work
    # Proves prefixing is a policy, not baked in.
    register_calculator(prefix: nil)

    assert @ctx.tools.key?("add")
    refute @ctx.tools.key?("calc__add")
  end

  def test_a_failing_server_raises_by_default
    # required: true is the default, and Tools::Mcp itself always raises —
    # the required/optional decision belongs to the caller, not here.
    assert_raises(StandardError) do
      Boukensha::Tools::Mcp.register(@registry, command: "/nonexistent/server")
    end

    assert_empty @ctx.tools
  end

  # ---------- the MUD server over MCP ---------------------------------------

  def test_registers_every_advertised_tool
    register_mud

    assert_equal 26, @ctx.tools.length
    assert @ctx.tools.key?("#{PREFIX}__look")
    assert @ctx.tools.key?("#{PREFIX}__cast_spell")
  end

  def test_registered_tools_dispatch_through_the_registry
    register_mud

    text = @registry.dispatch("#{PREFIX}__move", { "direction" => "north" })

    # Behaviour assertion — must pass untouched. Only the label moved.
    assert_match(/You north/, text)
    assert_includes @mud.commands, "north"
  end

  def test_the_daemon_still_sees_bare_names
    # The prefix is client-side. The wire is unchanged.
    client = register_mud

    assert_includes client.tools.map { |t| t["name"] }, "look"
    refute_includes client.tools.map { |t| t["name"] }, "#{PREFIX}__look"
  end

  def test_schemas_carry_the_enums_from_the_server
    register_mud

    move = @ctx.tools["#{PREFIX}__move"]
    assert_equal MudManager::Primitives::DIRECTIONS, move.parameters[:direction][:enum]
    assert_match(/one of: north/, move.parameters[:direction][:description])
  end

  def test_tool_failure_returns_text_rather_than_raising
    register_mud

    text = @registry.dispatch("#{PREFIX}__move", { "direction" => "widdershins" })

    assert_match(/INVALID_ARGUMENTS/, text)
  end

  def test_one_login_serves_many_tool_calls
    register_mud

    @registry.dispatch("#{PREFIX}__look", {})
    @registry.dispatch("#{PREFIX}__move", { "direction" => "east" })
    @registry.dispatch("#{PREFIX}__check", { "kind" => "score" })

    assert_equal 1, @mud.logins
  end
end

File: week2_capable/test/test_tools_mcp.py
import os
import subprocess
import sys
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import boukensha  # noqa: E402
from boukensha.tools import mcp as tools_mcp  # noqa: E402

# week2_capable/test/ -> week2_capable -> repo root. Was parents[4] while this
# tree lived at week1_baseline/python/12_context/test/.
REPO_ROOT = Path(__file__).resolve().parents[2]
DAEMON = REPO_ROOT / "week0_explore" / "mud_manager" / "bin" / "mud-manager"
FAKE_MUD_LIB = REPO_ROOT / "week0_explore" / "mud_manager" / "lib"
CALCULATOR = Path(__file__).resolve().parent / "support" / "tiny_mcp_server.py"
PASSWORD = "swordfish"
PREFIX = boukensha.MUD_PREFIX


class FakeMudProcess:
    """The Ruby FakeMud, driven as a subprocess.

    Porting it to Python would mean maintaining two fake MUDs that must agree
    on the CircleMUD login dance — a second thing to keep in sync for no gain.
    It is a server on a socket; the language it is written in is invisible from
    here, which is the same argument that lets Python skip tools/mud.rb.
    """

    def __init__(self, password=PASSWORD):
        self.password = password
        self.port = None
        self._process = None

    def start(self):
        script = (
            "require 'mud_manager'; require 'mud_manager/fake_mud'; "
            f"m = MudManager::FakeMud.new(password: {self.password!r}).start; "
            "puts m.port; $stdout.flush; sleep"
        )
        self._process = subprocess.Popen(
            ["ruby", "-I", str(FAKE_MUD_LIB), "-e", script],
            stdout=subprocess.PIPE,
            text=True,
            bufsize=1,
        )
        self.port = int(self._process.stdout.readline().strip())
        return self

    def stop(self):
        if not self._process:
            return

        self._process.terminate()
        try:
            self._process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self._process.kill()
        # Ruby closes the pipe when the process is reaped; Python holds the
        # TextIOWrapper open and warns. Not a leak that matters in a test, but a
        # ResourceWarning per test is noise that hides real ones.
        if self._process.stdout:
            self._process.stdout.close()
        self._process = None


@unittest.skipUnless(DAEMON.exists(), f"daemon not found at {DAEMON}")
class TestToolsMcp(unittest.TestCase):
    """Boukensha as an MCP host.

    These spawn real MCP servers as subprocesses, so they exercise the actual
    thing: an agent framework that knows nothing about MUDs acquiring tools at
    runtime. No API key, no billing, no real MUD.
    """

    def setUp(self):
        self.mud = FakeMudProcess().start()
        self.ctx = boukensha.Context(task=boukensha.Player, system="")
        self.registry = boukensha.Registry(self.ctx)
        self.clients = []

    def tearDown(self):
        for client in self.clients:
            try:
                client.close()
            except Exception:
                pass
        self.mud.stop()

    def register_mud(self, prefix=PREFIX):
        client = tools_mcp.register(
            self.registry,
            command="ruby",
            args=[str(DAEMON), "--mcp"],
            env={
                "MUD_HOST": "127.0.0.1",
                "MUD_PORT": str(self.mud.port),
                "MUD_NAME": "Gandalf",
                "MUD_PASSWORD": PASSWORD,
                "BOUKENSHA_DIR": "",
            },
            prefix=prefix,
            label="mud",
        )
        self.clients.append(client)
        return client

    def register_calculator(self, prefix="calc"):
        client = tools_mcp.register(
            self.registry,
            command=sys.executable,
            args=[str(CALCULATOR)],
            prefix=prefix,
            label="calc",
        )
        self.clients.append(client)
        return client

    # ---------- generic host behaviour ------------------------------------

    def test_registers_a_non_mud_server(self):
        # The point of the whole port: tools/mcp.py has no MUD knowledge. Proven
        # by demonstration rather than assertion — if MUD assumptions creep in,
        # this fails.
        self.register_calculator()

        self.assertEqual(sorted(self.ctx.tools), ["calc__add", "calc__shout"])
        self.assertEqual("42.0", self.registry.dispatch("calc__add", {"a": 2, "b": 40}))
        self.assertEqual("HELLO", self.registry.dispatch("calc__shout", {"text": "hello"}))

    def test_two_servers_coexist(self):
        self.register_mud()
        self.register_calculator()

        self.assertIn(f"{PREFIX}__look", self.ctx.tools)
        self.assertIn("calc__add", self.ctx.tools)
        self.assertEqual(28, len(self.ctx.tools))  # 26 MUD + 2 calculator

    def test_collision_raises_and_names_the_server(self):
        self.register_calculator(prefix="dup")

        with self.assertRaises(tools_mcp.ToolCollisionError) as caught:
            self.register_calculator(prefix="dup")

        self.assertIn("dup__add", str(caught.exception))
        self.assertIn("prefix", str(caught.exception))

    def test_bare_names_still_work(self):
        # Proves prefixing is a policy, not baked in.
        self.register_calculator(prefix=None)

        self.assertIn("add", self.ctx.tools)
        self.assertNotIn("calc__add", self.ctx.tools)

    def test_a_failing_server_raises(self):
        with self.assertRaises(Exception):
            tools_mcp.register(self.registry, command="/nonexistent/server")

        self.assertEqual({}, self.ctx.tools)

    # ---------- the MUD server over MCP -----------------------------------

    def test_registers_every_advertised_tool(self):
        self.register_mud()

        self.assertEqual(26, len(self.ctx.tools))
        self.assertIn(f"{PREFIX}__look", self.ctx.tools)
        self.assertIn(f"{PREFIX}__cast_spell", self.ctx.tools)

    def test_registered_tools_dispatch_through_the_registry(self):
        text = None
        self.register_mud()

        text = self.registry.dispatch(f"{PREFIX}__move", {"direction": "north"})

        self.assertIn("You north", text)

    def test_the_daemon_still_sees_bare_names(self):
        # The prefix is client-side. The wire is unchanged.
        client = self.register_mud()
        names = [t["name"] for t in client.tools]

        self.assertIn("look", names)
        self.assertNotIn(f"{PREFIX}__look", names)

    def test_schemas_carry_the_enums_from_the_server(self):
        self.register_mud()

        move = self.ctx.tools[f"{PREFIX}__move"]
        self.assertEqual(
            ["north", "east", "south", "west", "up", "down"],
            move.parameters["direction"]["enum"],
        )
        self.assertIn("one of: north", move.parameters["direction"]["description"])

    def test_tool_failure_returns_text_rather_than_raising(self):
        self.register_mud()

        text = self.registry.dispatch(f"{PREFIX}__move", {"direction": "widdershins"})

        self.assertIn("INVALID_ARGUMENTS", text)


if __name__ == "__main__":
    unittest.main()
